src/workerd/api/sockets.h

```diff
@@ -18,16 +18,30 @@ struct SocketAddress {
 
 struct SocketOptions {
   jsg::Unimplemented tls; // TODO(later): TCP socket options need to be implemented.
-  JSG_STRUCT(tls);
+  bool allowHalfOpen;
+  JSG_STRUCT(tls, allowHalfOpen);
 };
 
 class Socket: public jsg::Object {
 public:
-  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readable, jsg::Ref<WritableStream> writable,
-      kj::Own<jsg::PromiseResolverPair<void>> close)
-      : readable(kj::mv(readable)), writable(kj::mv(writable)),
+  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readableParam, jsg::Ref<WritableStream> writable,
+      kj::Own<jsg::PromiseResolverPair<void>> close, kj::Promise<void> connDisconnPromise,
+      jsg::Optional<SocketOptions> options)
+      : readable(kj::mv(readableParam)), writable(kj::mv(writable)),
         closeFulfiller(kj::mv(close)),
-        closedPromise(kj::mv(closeFulfiller->promise)) { };
+        closedPromise(kj::mv(closeFulfiller->promise)),
+        writeDisconnectedPromise(nullptr),
+        options(kj::mv(options)) {
+    // Listen for abrupt disconnects and resolve the `closed` promise when it occurs.
+    writeDisconnectedPromise = connDisconnPromise.then([this]() {
+      closeFulfiller->resolver.resolve();
```

Comment:
`connDsconnPromise` is a kj::Promise... but `closeFulfiller->resolver` is a JS promise resolver. This likely needs to use something like `ioContext.awaitIo(kj::mv(connDisconnPromise)).then(js, ...` to be safe.